lib/std/Thread/Pool.zig

```diff
@@ -162,41 +291,78 @@ pub fn spawn(pool: *Pool, comptime func: anytype, args: anytype) !void {
 
     {
         pool.mutex.lock();
-        defer pool.mutex.unlock();
 
-        const closure = try pool.allocator.create(Closure);
+        const closure = pool.allocator.create(Closure) catch {
+            pool.mutex.unlock();
+            @call(.auto, func, args);
+            return;
+        };
         closure.* = .{
             .arguments = args,
             .pool = pool,
         };
 
         pool.run_queue.prepend(&closure.run_node);
+        pool.mutex.unlock();
     }
 
     // Notify waiting threads outside the lock to try and keep the critical section small.
     pool.cond.signal();
 }
 
 fn worker(pool: *Pool) void {
+    var trash_buf: [1]u8 = undefined;
+    var connection: ?std.posix.fd_t = null;
+    defer if (connection) |fd| std.posix.close(fd);
+
     pool.mutex.lock();
     defer pool.mutex.unlock();
 
     while (true) {
         while (pool.run_queue.popFirst()) |run_node| {
-            // Temporarily unlock the mutex in order to execute the run_node
+            // Temporarily unlock the mutex in order to execute the run_node.
             pool.mutex.unlock();
             defer pool.mutex.lock();
 
+            if (connection == null) switch (pool.job_server_options) {
+                .abstain => {},
+                .connect, .host => |addr| lock: {
+                    const sockfd = std.posix.socket(
+                        std.posix.AF.UNIX,
+                        std.posix.SOCK.STREAM | std.posix.SOCK.CLOEXEC,
+                        0,
+                    ) catch |err| {
+                        std.log.debug("failed to make socket: {s}", .{@errorName(err)});
+                        break :lock;
+                    };
+                    connection = sockfd;
+
+                    std.posix.connect(sockfd, &addr.any, addr.getOsSockLen()) catch |err| {
+                        std.log.debug("failed to connect: {s}", .{@errorName(err)});
+                        break :lock;
+                    };
+
+                    _ = std.posix.read(sockfd, &trash_buf) catch |err| {
+                        std.log.debug("failed to read: {s}", .{@errorName(err)});
+                        break :lock;
+                    };
+                },
+            };
+
             const runFn = run_node.data.runFn;
             runFn(&run_node.data);
         }
 
         // Stop executing instead of waiting if the thread pool is no longer running.
-        if (pool.is_running) {
-            pool.cond.wait(&pool.mutex);
-        } else {
+        if (pool.end_flag)
             break;
+
+        if (connection) |fd| {
+            std.posix.close(fd);
+            connection = null;
```

Comment:
Feels like we might want to close connection _before_ we break out of the loop on the previous line?

Reply:
Ahh, nevermind, that's handler by the top-level defer.